### PR TITLE
Support @overload & @sharedRoute

### PR DESCRIPTION
**Pull request by MaryGao**
fixes https://github.com/Azure/autorest.typescript/issues/1760 https://github.com/Azure/autorest.typescript/issues/1763

Leverage the TypeScript overload to support the decorators @sharedRoute and @overload. 

Adding below test cases:
- @sharedRoute for different positions including query, body, header etc
- @overload for the same route, different routes and different actions
- different media types supporting by union, overload and sharedRoute

### CODE REVIEW

File: packages/typespec-ts/test/integration/typespec/overload/routes.tsp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using TypeSpec.Http;
+
+namespace TypeScript.Test.Overalod;
+
+@doc("Resource to process.")
+model Resource {
+  id: string;
+}
+@route("/overload")
+interface OveralodOperations {
```
Comment (MaryGao) [inline packages/typespec-ts/test/integration/typespec/overload/routes.tsp:12]: @joheredi Could you help review the overload cases and their generated code?

File: packages/typespec-ts/src/operationUtil.ts [hunk 2]
```diff
@@ -74,11 +75,15 @@ export function isDefinedStatusCode(statusCode: StatusCode) {
   return statusCode !== "*";
 }
 
-export function isBinaryPayload(body: Type, contentType: string) {
+export function isBinaryPayload(
```
Comment (joheredi) [inline packages/typespec-ts/src/operationUtil.ts:78]: Now that TypeSpec supports [@encode ](https://microsoft.github.io/typespec/standard-library/built-in-decorators#@encode) I wonder if we can leverage it to get rid of this complex logic.

File: packages/typespec-ts/test/integration/typespec/overload/routes.tsp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * Overload with same routes
+   */
+  @route("/get")
+  getThing(@body param: string | int32): string | int32;
+
+  @overload(OveralodOperations.getThing)
+  getString(@body param: string): string;
+
+  @overload(OveralodOperations.getThing)
```
Comment (joheredi) [inline packages/typespec-ts/test/integration/typespec/overload/routes.tsp:22]: @m-nash, did we get to an agreement on how we'll do overloads? Does this align with that?

**Review by joheredi — APPROVED** [review]

File: packages/typespec-ts/src/operationUtil.ts (hunk 2 above)
Comment (MaryGao) [reply packages/typespec-ts/src/operationUtil.ts:78]: https://github.com/Azure/autorest.typescript/blob/main/packages/typespec-ts/src/modelUtils.ts#L837-L847 Currently we already have this implemented.

This check helps on binary body not byte in a normal model.

File: packages/typespec-ts/test/integration/typespec/overload/routes.tsp (hunk 3 above)
Comment (m-nash) [reply packages/typespec-ts/test/integration/typespec/overload/routes.tsp:22]: I believe we agreed we would use sharedRoute in the main.tsp and remove the @overload decorator from azure.core and potentially add it for client.tsp only.

@srnagar can you confirm if that is correct?

File: packages/typespec-ts/test/integration/generated/mediaTypes/tspconfig.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  "@azure-tools/typespec-ts":
+    "emitter-output-dir": "{project-root}"
+    generateMetadata: false
+    generateTest: false
+    addCredentials: false
+    azureSdkForJs: false
+    isCadlTest: true
+    title: MediaTypesClient
+    packageDetails:
+      name: "@msinternal/media-types"
```
Comment (qiaozha) [inline packages/typespec-ts/test/integration/generated/mediaTypes/tspconfig.yaml:13]: Can you add create client test for these three cases ? just to make sure they can build successfully

**Review by qiaozha — COMMENTED** [review]
Do you mentioned that you will use the autorest testserver to write test cases for media types ?

**Comment by MaryGao** [comment]
> Do you mentioned that you will use the autorest testserver to write test cases for media types ?

I re-organizated these test cases and their path can't be shared.

File: packages/typespec-ts/test/integration/typespec/overload/routes.tsp (hunk 3 above)
Comment (MaryGao) [reply packages/typespec-ts/test/integration/typespec/overload/routes.tsp]: I moved the current overload cases to client.tsp. I notice that we have an issue https://github.com/Azure/typespec-azure/issues/2738 in Project TOM but not started yet.

Comment (srnagar) [reply packages/typespec-ts/test/integration/typespec/overload/routes.tsp]: > I believe we agreed we would use sharedRoute in the main.tsp and remove the @overload decorator from azure.core and potentially add it for client.tsp only.
> 
> @srnagar can you confirm if that is correct?

Yes, that is correct.

**Review by qiaozha — APPROVED** [review]

**Pull request merged**